### PR TITLE
[gh-2211] print nostr bech32 address.

### LINKED ISSUE #2211: Rooch CLI output the nostr address format: `npub1xxxx`

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

Rooch CLI `account list` command outputs the nostr bech32 public key.

- Closes #2211

### CODE REVIEW

File: crates/rooch-types/src/address.rs [hunk 1]
```diff
@@ -810,19 +811,29 @@ impl TryFrom<MultiChainAddress> for BitcoinAddress {
...
 }
 
+// TODO: rename NostrAddress to NostrPublicKey? https://github.com/nostr-protocol/nips/blob/master/19.md
 /// Nostr address type
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct NostrAddress(pub XOnlyPublicKey);
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct NostrAddress {
+    pub hex: XOnlyPublicKey,
+    pub bech32: String,
```
Comment (jolestar) [inline crates/rooch-types/src/address.rs:819]: We can just keep the XOnlyPublicKey field in NostrAddress, and output bech32 in the display function.

File: crates/rooch/src/commands/account/commands/list.rs [hunk 2]
```diff
@@ -45,6 +47,8 @@ impl LocalAccountView {
                 .bitcoin_address
                 .format(btc_network)
                 .expect("Failed to format bitcoin address"),
+            nostr_hex_address: account.nostr_address.hex.to_string(),
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/list.rs:50]: We do not need the nostr_hex_address

**Comment by jolestar** [comment]
After #2200 is merged, we can try to import the nostr private key and interact with rooch.

File: crates/rooch-types/src/address.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch-types/src/address.rs]: Resolved but there needs to deal with the test function to match hex and bech32 addresses.

File: crates/rooch/src/commands/account/commands/list.rs (hunk 2 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/list.rs]: > We do not need the nostr_hex_address

Removed.

File: crates/rooch-types/src/address.rs [hunk 3]
```diff
@@ -284,6 +286,26 @@ impl RoochAddress {
     }
 }
 
+pub trait Bech32Decode {
+    const LENGTH: usize;
+
+    fn hrp() -> &'static str;
+    fn decode(bech32: &str) -> Result<(String, Vec<u8>)>;
```
Comment (jolestar) [inline crates/rooch-types/src/address.rs:293]: 把 `from_bech32` 也放到这个 trait 里默认实现？
不然 `RoochAddress::from_bech32::<RoochAddress>(s)` 这样的调用很奇怪。

Comment (feliciss) [reply crates/rooch-types/src/address.rs]: > 把 `from_bech32` 也放到这个 trait 里默认实现？ 不然 `RoochAddress::from_bech32::<RoochAddress>(s)` 这样的调用很奇怪。

Yes. You're exactly correct on that. Thanks for pointing this out and I could resolve this issue as `RoochAddress::from_bech32` and `NostrPublicKey::from_bech32` without the trailing trait types.

**Review by jolestar — APPROVED** [review]

**Pull request merged**